### PR TITLE
refactor: rewrite @sapphire/result from scratch

### PR DESCRIPTION
**Pull request by kyranet**
Renames `Maybe<T>` to `Option<T>`, to align with Rust's names.

Also converted both `Option<T>` and `Result<T, E>` into classes with all the methods from [`std::option::Option<T>`](https://doc.rust-lang.org/std/option/enum.Option.html) and [`std::result::Result<T, E>`](https://doc.rust-lang.org/std/result/enum.Result.html) that could be ported to TypeScript.

Namespaced `ok`, `err`, `none`, `some`... they're now within the "enum" they are from, instead of being all over the place.

Made `Option.from` (previously `maybe()`) a lot more complete, aligning with what `Result.from` does. Also added `Option.fromAsync`.

# TODO

- [x] Make `Result.from` take Option.Resolvable<T> instead of only callback.
- [x] Make `Result.fromAsync` take Option.Resolvable<T> instead of only callback.
- [x] \*Lots* of tests.

# Squash Commit Body

```md
BREAKING CHANGE: Removed `Maybe` type, the substitute is `Option`
BREAKING CHANGE: Removed `maybe` function, the substitute is `Option.from`
BREAKING CHANGE: Removed `some` function, the substitute is `Option.some`
BREAKING CHANGE: Removed `none` function, the substitute is `Option.none`
BREAKING CHANGE: Removed `isSome` function, the substitute is `option.isSome`
BREAKING CHANGE: Removed `isNone` function, the substitute is `option.isNone`
BREAKING CHANGE: Removed `isMaybe` function, the substitute is `Option.is`
BREAKING CHANGE: Removed `UnwrapMaybeValue` type, the substitute is `Option.UnwrapSome`
BREAKING CHANGE: Removed `None` type, the substitute is `Option.None`
BREAKING CHANGE: Removed `Some` type, the substitute is `Option.Some`
BREAKING CHANGE: Removed `Err` type, the substitute is `Result.Err`
BREAKING CHANGE: Removed `Ok` type, the substitute is `Result.Ok`
BREAKING CHANGE: Removed `from` function, the substitute is `Result.from`
BREAKING CHANGE: Removed `fromAsync` function, the substitute is `Result.fromAsync`
```

### CODE REVIEW

File: packages/result/src/lib/Option/IOption.ts [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+	 * ```
+	 * @example
+	 * ```typescript
+	 * const x: Option<string> = none;
+	 * assert.equal(x.mapOrElse(() => 42, (value) => value.length), 42);
+	 * ```
+	 *
+	 * @see {@link https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or_else}
+	 */
+	mapOrElse<U>(defaultValue: () => U, cb: (value: T) => U): U;
```
[the remark this answers is not in the document]
Comment (kyranet) [reply packages/result/src/lib/Option/IOption.ts:204]: It's following the order in Rust's std, as seen in the link above.

**Review by vladfrangu — DISMISSED** [review]

File: packages/result/tests/common/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
+import { toString } from '../../src/lib/common/utils';
+
+describe('utils', () => {
+	describe('isFunction', () => {
+		// TODO: Write tests
```
Comment (favna) [inline packages/result/tests/common/utils.test.ts:5]: So do it then

File: packages/result/tests/common/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	});
+
+	describe('isPromise', () => {
+		// TODO: Write tests
```
Comment (favna) [inline packages/result/tests/common/utils.test.ts:9]: So do it then

File: packages/result/tests/common/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+			expect<typeof expected>(toString(given)).toBe(expected);
+			expect(fn).toHaveBeenCalledTimes(1);
+			expect(fn).toHaveBeenCalledWith();
+			expect(fn).toHaveReturnedWith(expected);
+		});
+	});
+
+	describe('valueOf', () => {
+		// TODO: Write tests
```
Comment (favna) [inline packages/result/tests/common/utils.test.ts:85]: So do it then

File: packages/result/tests/common/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	});
+
+	describe('toJSON', () => {
+		// TODO: Write tests
```
Comment (favna) [inline packages/result/tests/common/utils.test.ts:89]: So do it then

File: packages/result/tests/common/utils.test.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	});
+
+	describe('toPrimitive', () => {
+		// TODO: Write tests
```
Comment (favna) [inline packages/result/tests/common/utils.test.ts:93]: So do it then

File: packages/result/src/lib/Maybe.ts [hunk 7]
```diff
@@ -1,92 +0,0 @@
-import type { None, Some } from './Lexure';
-
-/**
- * A type used to express a value that may or may not exist.
- * @typeparam T The value's type.
- */
-export type Maybe<T> = Some<T> | None;
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:7]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `Maybe` type, the substitute is <something here>
```

File: packages/result/src/lib/Maybe.ts [hunk 8]
```diff
@@ -1,92 +0,0 @@
...
-
-/**
- * Returns the maybe itself.
- * @param value The value to convert.
- */
-export function maybe<T, V extends Maybe<T>>(value: V): V;
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:13]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `maybe` function, the substitute is <something here>
```

File: packages/result/src/lib/Maybe.ts [hunk 9]
```diff
@@ -1,92 +0,0 @@
...
-export function maybe<T>(value: T | Some<T>): Some<T>;
-export function maybe<T>(value: T | Maybe<T> | null): Maybe<T> {
-	return isMaybe(value) ? value : value === null ? none() : some(value);
-}
-
-/**
- * Creates a None with no value.
- * @return An existing Maybe.
- */
-export function some(): Some<unknown>;
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:40]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `some` function, the substitute is <something here>
```

File: packages/result/src/lib/Maybe.ts [hunk 10]
```diff
@@ -1,92 +0,0 @@
...
-export function some<T>(x: T): Some<T>;
-export function some<T>(x?: T): Some<unknown> {
-	return { exists: true, value: x };
-}
-
-/**
- * Creates a None value.
- * @return A non-existing Maybe.
- */
-export function none(): None {
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:57]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `none` function, the substitute is <something here>
```

File: packages/result/src/lib/Maybe.ts [hunk 11]
```diff
@@ -1,92 +0,0 @@
...
-	return { exists: false };
-}
-
-/**
- * Determines whether or not a Maybe is a Some.
- * @typeparam T The value's type.
- */
-export function isSome<T>(x: Maybe<T>): x is Some<T> {
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:65]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `isSome` function, the substitute is <something here>
```

File: packages/result/src/lib/Maybe.ts [hunk 12]
```diff
@@ -1,92 +0,0 @@
...
-	return x.exists;
-}
-
-/**
- * Determines whether or not a Maybe is a None.
- * @typeparam T The value's type.
- */
-export function isNone<T>(x: Maybe<T>): x is None {
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:73]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `isNone` function, the substitute is <something here>
```

File: packages/result/src/lib/Maybe.ts [hunk 13]
```diff
@@ -1,92 +0,0 @@
...
-	return !x.exists;
-}
-
-/**
- * Type-safe helper to preserve the type parameter's type.
- * @param x The value to check.
- */
-export function isMaybe<T>(x: Maybe<T>): true;
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:81]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `isMaybe` function, the substitute is <something here>
```

File: packages/result/src/lib/Maybe.ts [hunk 14]
```diff
@@ -1,92 +0,0 @@
...
-/**
- * Determines whether or not an arbitrary value is a Maybe.
- * @param x The value to check.
- */
-export function isMaybe<T>(x: unknown): x is Maybe<T>;
-export function isMaybe<T>(x: Maybe<T> | unknown): x is Maybe<T> {
-	return typeof x === 'object' && x !== null && typeof Reflect.get(x, 'exists') === 'boolean';
-}
-
-export type UnwrapMaybeValue<T extends Maybe<unknown>> = T extends Some<infer V> ? V : never;
```
Comment (favna) [inline packages/result/src/lib/Maybe.ts:92]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `UnwrapMaybeValue` type, the substitute is <something here>
```

File: packages/result/src/lib/Lexure.ts [hunk 15]
```diff
@@ -1,85 +0,0 @@
...
-	readonly value: T;
-}
-
-/**
- * @license MIT
- * @copyright 2020 1Computer
- *
- * The value does not exist.
- */
-export interface None {
```
Comment (favna) [inline packages/result/src/lib/Lexure.ts:78]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `None` type, the substitute is <something here>
```

File: packages/result/src/lib/Lexure.ts [hunk 16]
```diff
@@ -1,85 +0,0 @@
...
-export type Option<T> = Some<T> | None;
-
-/**
- * @license MIT
- * @copyright 2020 1Computer
- *
- * The value exists.
- * @typeparam T - Type of results.
- */
-export interface Some<T> {
```
Comment (favna) [inline packages/result/src/lib/Lexure.ts:60]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `Some` type, the substitute is <something here>
```

File: packages/result/src/lib/Lexure.ts [hunk 17]
```diff
@@ -1,85 +0,0 @@
...
-
-/**
- * @license MIT
- * @copyright 2020 1Computer
- *
- * A type that can express the lack of a value.
- * Used in this library for when a generic type could be nullable.
- * @typeparam T - Type of results.
- */
-export type Option<T> = Some<T> | None;
```
Comment (favna) [inline packages/result/src/lib/Lexure.ts:51]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `Option` type, the substitute is <something here>
```

File: packages/result/src/lib/Lexure.ts [hunk 18]
```diff
@@ -1,85 +0,0 @@
...
-}
-
-/**
- * @license MIT
- * @copyright 2020 1Computer
- *
- * The computation failed.
- * @typeparam E - Type of errors.
- */
-export interface Err<E> {
```
Comment (favna) [inline packages/result/src/lib/Lexure.ts:29]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `Err` type, the substitute is <something here>
```

File: packages/result/src/lib/Lexure.ts [hunk 19]
```diff
@@ -1,85 +0,0 @@
-/**
- * @license MIT
- * @copyright 2020 1Computer
- *
- * The computation is successful.
- * @typeparam T - Type of results.
- */
-export interface Ok<T> {
```
Comment (favna) [inline packages/result/src/lib/Lexure.ts:8]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `Ok` type, the substitute is <something here>
```

File: packages/result/src/lib/Parsers.ts [hunk 20]
```diff
@@ -1,34 +0,0 @@
-import { err, ok, Result } from './Result';
-
-/**
- * Creates a {@link Result} out of a callback.
- * @typeparam T The result's type.
- * @typeparam E The error's type.
- */
-export function from<T, E = unknown>(cb: (...args: unknown[]) => T): Result<T, E> {
```
Comment (favna) [inline packages/result/src/lib/Parsers.ts:8]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `from` function, the substitute is <something here>
```

File: packages/result/src/lib/Parsers.ts [hunk 21]
```diff
@@ -1,34 +0,0 @@
...
-		return err(error as E);
-	}
-}
-
-/**
- * Creates a {@link Result} out of a promise or async callback.
- * @typeparam T The result's type.
- * @typeparam E The error's type.
- */
-export async function fromAsync<T, E = unknown>(promiseOrCb: Awaitable<T> | ((...args: unknown[]) => Awaitable<T>)): Promise<Result<T, E>> {
```
Comment (favna) [inline packages/result/src/lib/Parsers.ts:21]: Document what to exchange this for in a 	`BREAKING CHANGE: ` commit line. I.e.:
```
BREAKING CHANGE: Removed `fromAsync` function, the substitute is <something here>
```

**Review by favna — CHANGES_REQUESTED** [review]

**Review by favna — CHANGES_REQUESTED** [review]
Needs more test coverage. See coverage report on codecov / local

**Review by favna — CHANGES_REQUESTED** [review]
Document breaking changes as mentioned above but also in general

File: .vscode/settings.json [hunk 22]
```diff
@@ -5,5 +5,6 @@
 	"editor.insertSpaces": false,
 	"editor.detectIndentation": false,
 	"files.eol": "\n",
-	"typescript.tsdk": "node_modules\\typescript\\lib"
+	"typescript.tsdk": "node_modules\\typescript\\lib",
```
Comment (favna) [inline .vscode/settings.json:8]:
```suggestion
```

VSCode is on TS 3.7 now so we can remove this line again
(it should've never been checked it tbh)

File: .vscode/settings.json [hunk 23]
```diff
@@ -5,5 +5,6 @@
...
+	"testing.automaticallyOpenPeekView": "never"
```
Comment (favna) [inline .vscode/settings.json:9]: What is this setting?

**Review by favna — CHANGES_REQUESTED** [review]

File: .vscode/settings.json (hunk 23 above)
Comment (kyranet) [reply .vscode/settings.json:9]: Uhhhhhhhhhhhhhhhhhhhhhhhhhhhh

Something the Vitest extension added when I tried to run just the Result tests

**Review by favna — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**